include/metal/list/partition.hpp

```diff
@@ -36,8 +41,73 @@ namespace metal {
     /// ### See Also
     /// \see list, copy_if, remove_if
     template<class seq, class lbd>
-    using partition =
-        metal::pair<metal::copy_if<seq, lbd>, metal::remove_if<seq, lbd>>;
+    using partition = detail::call<detail::_partition<lbd>::template type, seq>;
+}
+
+#include "../lambda/lambda.hpp"
+#include "../list/join.hpp"
+#include "../list/list.hpp"
+#include "../number/if.hpp"
+#include "../number/number.hpp"
+#include "../pair/pair.hpp"
+
+namespace metal {
+    /// \cond
+    namespace detail {
+        template<class...>
+        struct _partition_joiner {};
+
+        template<class... left, class... right>
+        struct _partition_joiner<list<left, right>...> {
+            using type = pair<join<left...>, join<right...>>;
+        };
+
+        template<bool cond>
+        struct _partition_filter {
+            template<class val>
+            using type = list<list<val>, list<>>;
+        };
+
+        template<>
+        struct _partition_filter<false> {
+            template<class val>
+            using type = list<list<>, list<val>>;
+        };
+
+        template<class conds, class seq>
+        struct _partitioner {};
+
+        template<int_... ns, class... vals>
+        struct _partitioner<list<number<ns>...>, list<vals...>>
+            : _partition_joiner<
+                  typename _partition_filter<ns>::template type<vals>...> {};
```

Comment:
Have we tested `if_<number<ns>, pair<list<val>, list<>>, pair<list<>, list<val>>>...` vs `_partition_filter`?

Reply:
> Have we tested `if_<number<ns>, pair<list<val>, list<>>, pair<list<>, list<val>>>...` vs `_partition_filter`?

I am not absolutely sure about that. I will benchmark it.

I remember why I once introduced that `filter`: What you propose here is similar to [this pattern](https://github.com/brunocodutra/metal/pull/115/commits/34eeae4d64dfe52e015b169b82ce166be2b9f0bb#diff-2a7fabd857009f87aacd803768e57228R59), which is the ["GCC9 killer (v0)"](https://ecrypa.github.io/metal-benchmark/nth_element_gcc9/). I thus introduced the indirection via [`_fast_partition_helper`](https://github.com/brunocodutra/metal/pull/115/commits/34eeae4d64dfe52e015b169b82ce166be2b9f0bb#diff-d77f2ed31417a4569b562d1540f6dbcbR21) in v2, which is faster on GCC9. When extracting the partitioner, I probably renamed `_fast_partition_helper` into `filter`.

Reply:
I'm happy to merge this as is either way and iterate on it later if necessary, unless you have plans for this PR still?

Reply:
Good suggestion! Feel free to merge.